File: src/main/scala/com/exasol/cloudetl/emitter/FilesMetadataEmitter.scala
package com.exasol.cloudetl.emitter

import java.lang.Long

import scala.collection.mutable.ListBuffer

import com.exasol.ExaIterator
import com.exasol.cloudetl.bucket.Bucket
import com.exasol.cloudetl.storage.FileFormat
import com.exasol.cloudetl.storage.StorageProperties
import com.exasol.parquetio.splitter.ParquetFileSplitter

import com.typesafe.scalalogging.LazyLogging
import org.apache.parquet.hadoop.util.HadoopInputFile

/**
 * A class that calculates and emits file based metadata from a storage path.
 *
 * @param properties a storage properties to create configurations
 * @param parallelism a parallelism number to create groups
 */
final case class FilesMetadataEmitter(properties: StorageProperties, parallelism: Int)
    extends Emitter
    with LazyLogging {

  private[this] val bucket = Bucket(properties)
  private[this] val paths = bucket.getPaths()
  private[this] val fileFormat = properties.getFileFormat()

  override def emit(context: ExaIterator): Unit = {
    logger.info(s"Found total of '${paths.size}' files in path '${bucket.bucketPath}'.")
    if (fileFormat != FileFormat.PARQUET) {
      emitRegularFilesMetadata(context)
    } else {
      emitParquetFilesMetadata(context)
    }
  }

  private[this] def emitRegularFilesMetadata(context: ExaIterator): Unit = {
    var index = 0L
    paths.foreach { case filename =>
      context.emit(filename.toString(), s"${index % parallelism}", Long.valueOf(0), Long.valueOf(0))
      index += 1L
    }
  }

  private[this] def emitParquetFilesMetadata(context: ExaIterator): Unit = {
    val chunkSize = properties.getChunkSize()
    val chunks = ListBuffer.empty[FilenameChunkInterval]
    paths.foreach { case filename =>
      val inputFile = HadoopInputFile.fromPath(filename, bucket.getConfiguration())
      val splits = new ParquetFileSplitter(inputFile, chunkSize).getSplits()
      for { i <- 0 until splits.size() } {
        chunks.append(
          FilenameChunkInterval(filename.toString(), splits.get(i).getStartPosition(), splits.get(i).getEndPosition())
        )
      }
    }
    val partitioner = Partitioner(chunks.size, parallelism)
    var index = 0L
    var count = 0L
    chunks.foreach { case interval =>
      count += 1
      context.emit(interval.filename, s"$index", interval.start, interval.end)
      if (partitioner.isNewPartition(count)) {
        index += 1
      }
    }
  }

  private[this] case class FilenameChunkInterval(filename: String, start: Long, end: Long)

  private[this] case class Partitioner(total: Int, requestedPartitionSize: Int) {
    val numberOfPartitions: Long = math.ceil(total / requestedPartitionSize.toDouble).toLong
    val partitionSize: Long = math.floor(total / numberOfPartitions.toDouble).toLong
    var leftOvers: Long = total % numberOfPartitions

    def isNewPartition(currentCount: Long): Boolean =
      if (leftOvers > 0 && currentCount >= partitionSize + 1) {
        leftOvers -= 1
        true
      } else if (leftOvers == 0 && currentCount >= partitionSize) {
        true
      } else {
        false
      }
  }

}

File: src/main/java/com/exasol/cloudetl/emitter/FilesMetadataEmitter.java
package com.exasol.cloudetl.emitter;

import java.util.ArrayList;
import java.util.List;

import org.apache.parquet.hadoop.util.HadoopInputFile;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.exasol.ExaIterator;
import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.bucket.Bucket;
import com.exasol.cloudetl.storage.FileFormat;
import com.exasol.cloudetl.storage.StorageProperties;
import com.exasol.parquetio.splitter.ParquetFileSplitter;

/** Emits file metadata into Exasol. */
public final class FilesMetadataEmitter implements Emitter {
    private static final Logger LOGGER = LoggerFactory.getLogger(FilesMetadataEmitter.class);

    private final StorageProperties properties;
    private final int parallelism;
    private final Bucket bucket;
    private final scala.collection.immutable.Seq<org.apache.hadoop.fs.Path> paths;
    private final FileFormat fileFormat;

    /** Create an emitter. */
    public FilesMetadataEmitter(final StorageProperties properties, final int parallelism) {
        this.properties = properties;
        this.parallelism = parallelism;
        this.bucket = Bucket.create(properties);
        this.paths = this.bucket.getPaths();
        this.fileFormat = properties.getFileFormat();
    }

    @Override
    public void emit(final ExaIterator context) {
        LOGGER.info("Found total of '{}' files in path '{}'.", this.paths.size(), this.bucket.bucketPath());
        if (isParquetFormat(this.fileFormat)) {
            emitParquetFilesMetadata(context);
        } else {
            emitRegularFilesMetadata(context);
        }
    }

    private boolean isParquetFormat(final FileFormat fileFormat) {
        return fileFormat == FileFormat.PARQUET || fileFormat == FileFormat.DELTA;
    }

    private void emitRegularFilesMetadata(final ExaIterator context) {
        long index = 0L;
        for (final org.apache.hadoop.fs.Path filename : ScalaConverters.asJavaList(this.paths)) {
            emitRow(context, filename.toString(), String.valueOf(index % this.parallelism), Long.valueOf(0), Long.valueOf(0));
            index++;
        }
    }

    private void emitParquetFilesMetadata(final ExaIterator context) {
        final long chunkSize = this.properties.getChunkSize();
        final List<FilenameChunkInterval> chunks = new ArrayList<>();
        try {
            for (final org.apache.hadoop.fs.Path filename : ScalaConverters.asJavaList(this.paths)) {
                final HadoopInputFile inputFile = HadoopInputFile.fromPath(filename, this.bucket.getConfiguration());
                final java.util.List<com.exasol.parquetio.data.ChunkInterval> splits =
                        new ParquetFileSplitter(inputFile, chunkSize).getSplits();
                for (int index = 0; index < splits.size(); index++) {
                    chunks.add(new FilenameChunkInterval(filename.toString(), splits.get(index).getStartPosition(),
                            splits.get(index).getEndPosition()));
                }
            }
        } catch (final java.io.IOException exception) {
            throw new IllegalStateException(exception);
        }
        final Partitioner partitioner = new Partitioner(chunks.size(), this.parallelism);
        long index = 0L;
        long count = 0L;
        for (final FilenameChunkInterval interval : chunks) {
            LOGGER.info("Emitting filename metadata {} -> [{} ... {}) on index {}.", interval.filename, interval.start,
                    interval.end, index);
            emitRow(context, interval.filename, String.valueOf(index), interval.start, interval.end);
            count++;
            if (partitioner.isNewPartition(count)) {
                index++;
                count = 0;
            }
        }
    }

    private static final class FilenameChunkInterval {
        private final String filename;
        private final long start;
        private final long end;

        private FilenameChunkInterval(final String filename, final long start, final long end) {
            this.filename = filename;
            this.start = start;
            this.end = end;
        }
    }

    private static final class Partitioner {
        private final long partitionSize;
        private long leftOvers;

        private Partitioner(final int total, final int numberOfPartitions) {
            this.partitionSize = (long) Math.floor(total / (double) numberOfPartitions);
            this.leftOvers = total % numberOfPartitions;
        }

        private boolean isNewPartition(final long currentCount) {
            if (this.leftOvers > 0 && currentCount >= this.partitionSize + 1) {
                this.leftOvers--;
                return true;
            } else if (this.leftOvers == 0 && currentCount >= this.partitionSize) {
                return true;
            }
            return false;
        }
    }

    private static void emitRow(final ExaIterator context, final Object... values) {
        try {
            context.emit(values);
        } catch (final Exception exception) {
            throw new IllegalStateException(exception);
        }
    }
}
